### DISCUSSION TITLE
Namespace package template

### DISCUSSION THREAD

**Discussion by kristang** [copier-org/copier#845] [Q&A] [answered]
Hello, I am having trouble trying to create template for namespace packages.

I am on the following system:

- **copier**:  `7.0.1`
- **Python**: `3.9.13`
- **OS**: `Windows 11 Enterprise (build 22000.856)`

It seems sort of related to the discussion in https://github.com/copier-org/copier/discussions/683 and this might just be because I am a new user and am failing to understand the placement of different files or where to invoke the commands from.

I want to create a folder structure where the top-level directory is generated as `"{{namespace}}-{{module_name}}"` with subfolder `"{{namespace}}"` which then contains a subfolder itself `"{{module_name}}"`

I have tried different things, and the closest I have been able to achieve what I want is a template structure as follows:

```
📁 namespace-package-copier 
├── 📁 .git
├── 📁 project
    └──📁 {{namespace}}-{{module_name}}
       └──📁 {{namespace}} 
          └──📁 {{module_name}} 
             └──📄 __main__.py.jinja
       ├──📄 .gitignore
       ├──📄 README.md
       └──📄 {{_copier_conf.answers_file}}.jinja
├──📄 copier.yml
├──📄 .gitignore
└──📄 README.md
```

The copier.yml looks like this:

```yaml
namespace:
    type: str
    help: What is your project namespace?

module_name:
    type: str
    help: What is your Python module name?

_subdirectory: project
```

Envoking this:
```bash
λ copier ../namespace-package-copier copier-testing/.
No git tags found in template; using HEAD as ref
� What is your project namespace?
   placeholder (this is pre-filled so it seems copier correctly captures existing answers)
� What is your Python module name?
   module (this is pre-filled so it seems copier correctly captures existing answers)

Copying from template version 0.0.0.post43.dev0+b639d5b
 identical  .
    create  placeholder-module
    create  placeholder-module\.gitignore
    create  placeholder-module\README.md
    create  placeholder-module\placeholder
    create  placeholder-module\placeholder\module
    create  placeholder-module\placeholder\module\__main__.py
    create  placeholder-module\.copier-answers.yml
```

Will yield a new directory inside the `packages` directory called `placeholder-module` with the following structure:

```
📁 placeholder-module
├──📁 .git
└──📁 placeholder
   └──📁 module 
      └──📄 __main__.py
├──📄 .gitignore
├──📄 README.md
└──📄 .copier-answers.yml
```
Git initializing this project and then attempting to update it breaks however. 

When making a change in the template and running `copier update` the project is not updated, but the project recreated inside the existing project:

```bash
λ copier update copier-testing/placeholder-module/
No git tags found in template; using HEAD as ref
Updating to template version 0.0.0.post44.dev0+89a5a5c
No git tags found in template; using HEAD as ref
� What is your project namespace?
   placeholder
� What is your Python module name?
   module

Copying from template version 0.0.0.post44.dev0+89a5a5c
 identical  .
    create  placeholder-module
    create  placeholder-module\.gitignore
    create  placeholder-module\README.md
    create  placeholder-module\placeholder
    create  placeholder-module\placeholder\module
    create  placeholder-module\placeholder\module\__main__.py
    create  placeholder-module\.copier-answers.yml
```

Yielding:
```
📁 placeholder-module
├──📁 .git
└──📁 placeholder
   └──📁 module 
      └──📄 __main__.py
├──📄 .gitignore
├──📄 README.md
├──📄 .copier-answers.yml ------ contains: _commit: b639d5b
└──📁 placeholder-module
   └──📁 placeholder
      └──📁 module 
         └──📄 __main__.py
   ├──📄 .gitignore
   ├──📄 README.md
   └──📄 .copier-answers.yml ------ contains: _commit: 89a5a5c
```

Where am I going wrong?

**Comment by pawamoy** [comment, marked as answer]
You are generating the project with `copier-testing` as destination. Then you are updating `copier-testing/placeholder-module`. Shouldn't you update `copier-testing` directly?

**Reply by kristang** [reply]
That would require the `copier-testing` directory to be a git-tracked repo. I am interested in git tracking the `placeholder-module` directory.

**Reply by kristang** [reply]
Updated original post with .git folders to accurately reflect git tracking

**Reply by pawamoy** [reply]
Hmmmm. I don't think that is supported. The most simple alternative I see is that you remove the first layer of directories and pass the right name when generating the project:

```bash
📁 namespace-package-copier 
├── 📁 .git
├── 📁 project
     └──📁 {{namespace}} 
        └──📁 {{module_name}} 
           └──📄 __main__.py.jinja
     ├──📄 .gitignore
     ├──📄 README.md
     └──📄 {{_copier_conf.answers_file}}.jinja
├──📄 copier.yml
├──📄 .gitignore
└──📄 README.md
```

```bash
# generate
copier ../namespace-package-copier copier-testing/mynamespace-mymodule
# update
copier update copier-testing/mynamespace-mymodule
```

What would be nice is the ability to dynamically change the output directory name based on the answers. Maybe that's doable with a task.

**Reply by kristang** [reply]
Yeah, the entire point of my exercise here was I don't want to leave it to the users what to name that top-level folder when invoking copier. It should always follow `namespace-module` convention. 

The other thing I tried was discarding `_subdirectory` in `copier.yml` and having everything in the same folder - but that copies the `.gitignore`and `README.md` from the toplevel directory, which are supposed to reflect the copier template itself and should not be copied along to any new projects generated from the template, as well as stores answers outside of the directory I actually plan on git tracking.

```
📁 namespace-package-copier 
├── 📁 .git
└── 📁 {{namespace}}-{{module_name}}
     └──📁 {{namespace}} 
        └──📁 {{module_name}} 
           └──📄 __main__.py.jinja
     ├──📄 .gitignore
     └──📄 README.md
├──📄 {{_copier_conf.answers_file}}.jinja
├──📄 copier.yml
├──📄 .gitignore
└──📄 README.md
``` 

Which then yields:
```
📁 placeholder-module
  └──📁 placeholder
     └──📁 module
        └──📄 __main__.py.jinja
  ├──📄 .gitignore
  └──📄 README.md
├──📄 .copier-answers.yml
├──📄 .gitignore
└──📄 README.md
``` 

Maybe just ignoring everything in the top-level directory might solve my problem, except `_exclude = - "README.md"` in `copier.yml` seems to exclude README in all directories.

**Reply by yajo** [reply]
It is a strict requirement that a project is git-tracked, for it to be updated. That's the only way to know how it was before the update.

If your don't want that, then you cannot use updates. 🤷‍♂️

**Reply by kristang** [reply]
I understand that the project needs to be git-tracked. I just assumed that the entry point for updates would always be the root-folder where `.copier-answers.yml` is located, but I can see how that would break if the top-level directory was created by `copier`

Maybe a task would solve this.
